### PR TITLE
Add possibility to specify further attributes for SAML mapping

### LINKED ISSUE #12: Pass attributes through to SAML
privacyidea itself allows a resolver to return further attributes (for example, groups). However, the simplesamlphp module is limited to the attributes it knows about [privacyidea.php#L236](/privacyidea/simplesamlphp-module-privacyidea/blob/master/lib/Auth/Source/privacyidea.php#L236). 

Could this be extended by a configuration option or be changed to map all attributes? (Personally, I just added my own attribute to the list)

### PR DESCRIPTION
**Pull request by andyboeh**
This allows to specify further attributes on the attributemap configuration. Only attributes configured here are added to the SAML response, all others are left untouched. Closes #12
<a href='#crh-start'></a><a href='#crh-data-%7B%22processed%22%3A%20%5B%22https%3A//github.com/privacyidea/simplesamlphp-module-privacyidea/pull/14%23discussion_r94240347%22%2C%20%22https%3A//github.com/privacyidea/simplesamlphp-module-privacyidea/pull/14%23issuecomment-269791496%22%2C%20%22https%3A//github.com/privacyidea/simplesamlphp-module-privacyidea/pull/14%23discussion_r94251482%22%2C%20%22https%3A//github.com/privacyidea/simplesamlphp-module-privacyidea/pull/14%23discussion_r94270716%22%2C%20%22https%3A//github.com/privacyidea/simplesamlphp-module-privacyidea/pull/14%23discussion_r94271438%22%2C%20%22https%3A//github.com/privacyidea/simplesamlphp-module-privacyidea/pull/14%23issuecomment-269857898%22%2C%20%22https%3A//github.com/privacyidea/simplesamlphp-module-privacyidea/pull/14%23issuecomment-269857987%22%2C%20%22https%3A//github.com/privacyidea/simplesamlphp-module-privacyidea/pull/14%23issuecomment-269946468%22%5D%2C%20%22comments%22%3A%20%7B%22General%20Comment%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/privacyidea/simplesamlphp-module-privacyidea/pull/14%23issuecomment-269791496%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Thanks%20a%20lot%20for%20the%20PR.%20A%20comment%20like%5Cr%5Cn%5Cr%5CnCloses%20%2312%20%5Cr%5Cn%5Cr%5CnWill%20automatically%20close%20the%20issue%20%2312%20on%20submit.%22%2C%20%22created_at%22%3A%20%222016-12-30T16%3A31%3A11Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1908620%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/cornelinux%22%7D%7D%2C%20%7B%22body%22%3A%20%22It%20seems%20that%20one%20can%27t%20close%20issues%20by%20mentioning%20it%20in%20the%20commit%20message%21%3F%20I%27m%20used%20to%20phabricator%2C%20where%20this%20works%20perfectly%20fine.%20%5Cr%5CnAnyhow%2C%20this%20should%20now%20close%20%2312.%22%2C%20%22created_at%22%3A%20%222016-12-31T09%3A49%3A13Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/3099753%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/andyboeh%22%7D%7D%2C%20%7B%22body%22%3A%20%22%3A%2B1%3A%22%2C%20%22created_at%22%3A%20%222016-12-31T09%3A52%3A38Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/3099753%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/andyboeh%22%7D%7D%2C%20%7B%22body%22%3A%20%22Thanks%20a%20lot%21%22%2C%20%22created_at%22%3A%20%222017-01-02T08%3A57%3A50Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1908620%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/cornelinux%22%7D%7D%5D%2C%20%22title%22%3A%20%22General%20Comment%22%7D%2C%20%22Pull%20d29f34f1eb5c5e02f24c358fd873d40e5e371e0d%20lib/Auth/Source/privacyidea.php%2014%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/privacyidea/simplesamlphp-module-privacyidea/pull/14%23discussion_r94240347%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22According%20to%20http%3A//php.net/manual/en/function.array-merge.php%20dublicate%20keys%20in%20the%20array%20will%20be%20overwritten.%20So%20I%20guess%20we%20can%20skip%20the%20%2Aarray_diff%2A%20and%20%20and%20just%20do%5Cr%5Cn%5Cr%5Cn%20%20%20%20%24arr%20%3D%20array_merge%28array_keys%28%24this-%3Eattributemap%29%2C%20%24arr%29%3B%5Cr%5Cn%5Cr%5CnTo%20my%20understanding%20there%20will%20be%20no%20dublicate%20values%3F%22%2C%20%22created_at%22%3A%20%222016-12-30T16%3A30%3A00Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1908620%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/cornelinux%22%7D%7D%2C%20%7B%22body%22%3A%20%22You%20are%20completely%20right%2C%20I%20did%20not%20read%20the%20documentation%20thoroughly%20enough.%20Plus%2C%20the%20documentation%20needs%20updating%20to%20reflect%20the%20changed%20behaviour.%20I%27ll%20make%20a%20new%20commit%20tomorrow.%22%2C%20%22created_at%22%3A%20%222016-12-30T19%3A16%3A48Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/3099753%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/andyboeh%22%7D%7D%2C%20%7B%22body%22%3A%20%22thx%22%2C%20%22created_at%22%3A%20%222016-12-31T08%3A11%3A38Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1908620%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/cornelinux%22%7D%7D%2C%20%7B%22body%22%3A%20%22%3A%2B1%3A%22%2C%20%22created_at%22%3A%20%222016-12-31T09%3A47%3A19Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/3099753%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/andyboeh%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20lib/Auth/Source/privacyidea.php%3AL236-244%22%7D%7D%7D'></a>
<a href='https://www.codereviewhub.com/'></a>
- [ ] <a href='#crh-comment-General Comment'></a> &nbsp;<b><a href='https://github.com/privacyidea/simplesamlphp-module-privacyidea/pull/14#issuecomment-269791496'>General Comment</a></b>
- <a href='https://github.com/cornelinux'></a> Thanks a lot for the PR. A comment like
Closes #12
Will automatically close the issue #12 on submit.
- <a href='https://github.com/andyboeh'></a> It seems that one can't close issues by mentioning it in the commit message!? I'm used to phabricator, where this works perfectly fine.
Anyhow, this should now close #12.
- <a href='https://github.com/andyboeh'></a> :+1:
- <a href='https://github.com/cornelinux'></a> Thanks a lot!
- [x] <a href='#crh-comment-Pull d29f34f1eb5c5e02f24c358fd873d40e5e371e0d lib/Auth/Source/privacyidea.php 14'></a> &nbsp;<b><a href='https://github.com/privacyidea/simplesamlphp-module-privacyidea/pull/14#discussion_r94240347'>File: lib/Auth/Source/privacyidea.php:L236-244</a></b>
- <a href='https://github.com/cornelinux'></a> According to http://php.net/manual/en/function.array-merge.php dublicate keys in the array will be overwritten. So I guess we can skip the *array_diff* and  and just do
$arr = array_merge(array_keys($this->attributemap), $arr);
To my understanding there will be no dublicate values?
- <a href='https://github.com/andyboeh'></a> You are completely right, I did not read the documentation thoroughly enough. Plus, the documentation needs updating to reflect the changed behaviour. I'll make a new commit tomorrow.
- <a href='https://github.com/cornelinux'></a> thx

<a href='https://www.codereviewhub.com/privacyidea/simplesamlphp-module-privacyidea/pull/14?mark_as_completed=1'></a>&nbsp;<a href='https://www.codereviewhub.com/privacyidea/simplesamlphp-module-privacyidea/pull/14?approve=1'></a>&nbsp;<a href='https://github.com/privacyidea/simplesamlphp-module-privacyidea/pull/14'></a>
<a href='#crh-end'></a>

### CODE REVIEW

**Comment by cornelinux** [comment]
This pull request is now managed with CodeReviewHub. <a href='https://github.com/privacyidea/simplesamlphp-module-privacyidea/pull/14'>Refresh</a>.<a href='#crh-update-none'></a>

File: lib/Auth/Source/privacyidea.php [hunk 1]
```diff
@@ -234,6 +236,8 @@ protected function login_chal_resp($username, $password, $transaction_id, $signa
         /* If we get this far, we have a valid login. */
         $attributes = array();
         $arr = array("username", "surname", "email", "givenname", "mobile", "phone", "realm", "resolver");
+        // Add all additional attributes defined in the array map to the search array
+        $arr = array_merge(array_diff(array_keys($this->attributemap), $arr), $arr);
```
Comment (cornelinux) [inline lib/Auth/Source/privacyidea.php]: According to http://php.net/manual/en/function.array-merge.php dublicate keys in the array will be overwritten. So I guess we can skip the *array_diff* and  and just do

    $arr = array_merge(array_keys($this->attributemap), $arr);

To my understanding there will be no dublicate values?

**Comment by cornelinux** [comment]
Thanks a lot for the PR. A comment like

Closes #12 

Will automatically close the issue #12 on submit.

File: lib/Auth/Source/privacyidea.php (hunk 1 above)
Comment (andyboeh) [inline lib/Auth/Source/privacyidea.php]: You are completely right, I did not read the documentation thoroughly enough. Plus, the documentation needs updating to reflect the changed behaviour. I'll make a new commit tomorrow.

Comment (cornelinux) [inline lib/Auth/Source/privacyidea.php]: thx

Comment (andyboeh) [inline lib/Auth/Source/privacyidea.php]: :+1:

**Comment by andyboeh** [comment]
It seems that one can't close issues by mentioning it in the commit message!? I'm used to phabricator, where this works perfectly fine. 
Anyhow, this should now close #12.

**Comment by andyboeh** [comment]
:+1:

**Comment by cornelinux** [comment]
Thanks a lot!

**Pull request merged**